iceoryx_hoofs/source/log/platform_building_blocks/console_logger.cpp

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/log/platform_building_blocks/console_logger.hpp"
+#include "iceoryx_hoofs/cxx/attributes.hpp"
+#include "iceoryx_hoofs/platform/time.hpp"
+
+#include <cstdio>
+#include <cstring>
+#include <ctime>
+
+namespace iox
+{
+namespace pbb
+{
+// NOLINTJUSTIFICATION see at declaration in header
+// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)
+std::atomic<LogLevel> ConsoleLogger::m_activeLogLevel{LogLevel::INFO};
+
+// NOLINTJUSTIFICATION see at declaration in header
+// NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+thread_local char ConsoleLogger::m_buffer[ConsoleLogger::NULL_TERMINATED_BUFFER_SIZE]{0};
+
+thread_local uint32_t ConsoleLogger::m_bufferWriteIndex{0U};
+// NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
+
+LogLevel ConsoleLogger::getLogLevel() noexcept
+{
+    return m_activeLogLevel.load(std::memory_order_relaxed);
+}
+
+// NOLINTJUSTIFICATION member access is required
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+void ConsoleLogger::setLogLevel(const LogLevel logLevel) noexcept
+{
+    m_activeLogLevel.store(logLevel, std::memory_order_relaxed);
+}
+
+void ConsoleLogger::createLogMessageHeader(const char* file,
+                                           const int line,
+                                           const char* function,
+                                           LogLevel logLevel) noexcept
+{
+    timespec timestamp{0, 0};
+    if (clock_gettime(CLOCK_REALTIME, &timestamp) != 0)
```

Comment:
Hey it is a me ... the party pooper :smile: 

What happens when the time changes? Then we have log messages from the future? And this is a real use case I think we have to support.

I would propose to use `CLOCK_MONOTONIC` were I have to admit that it is less readable but it provides at least a total order so that one knows what was printed when.

Another readable approach would an id in front of every log message which is incremented. Then we also know when what message was produced.

Reply:
@elfenpiff yes, but not for the MVP. This does what the current logger does. In the future it might be a compile time option/env variable to change the behavior.

Reply:
added in #1621

Reply:
@elBoberido can you please hint me where this was addressed in #1621 ? I see that you use `CLOCK_REALTIME` still there and also do not add an incrementing id which puts them into total order.
Did you maybe mean you added it to another issue?

Reply:
@elfenpiff upsi. This was actually meant for https://github.com/eclipse-iceoryx/iceoryx/pull/1569#discussion_r963793998. Seems I posted it to the wrong thread.

I added a task to #1345